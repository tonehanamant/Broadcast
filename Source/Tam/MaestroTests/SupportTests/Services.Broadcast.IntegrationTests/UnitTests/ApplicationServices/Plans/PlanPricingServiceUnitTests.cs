using Common.Services;
using Common.Services.Repositories;
using Hangfire;
using Moq;
using NUnit.Framework;
using Services.Broadcast.ApplicationServices;
using Services.Broadcast.BusinessEngines;
using Services.Broadcast.Cache;
using Services.Broadcast.Clients;
using Services.Broadcast.Entities.Enums;
using Services.Broadcast.Entities.Plan.Pricing;
using Services.Broadcast.Repositories;
using System;

namespace Services.Broadcast.IntegrationTests.UnitTests.ApplicationServices.Plans
{
    [TestFixture]
    public class PlanPricingServiceUnitTests
    {
        private readonly Mock<IDataRepositoryFactory> _DataRepositoryFactoryMock;
        private readonly Mock<ISpotLengthEngine> _SpotLengthEngineMock;
        private readonly Mock<IPricingApiClient> _PricingApiClientMock;
        private readonly Mock<IBackgroundJobClient> _BackgroundJobClientMock;
        private readonly Mock<IPlanPricingInventoryEngine> _PlanPricingInventoryEngineMock;
        private readonly Mock<IBroadcastLockingManagerApplicationService> _BroadcastLockingManagerApplicationServiceMock;

        public PlanPricingServiceUnitTests()
        {
            _DataRepositoryFactoryMock = new Mock<IDataRepositoryFactory>();
            _SpotLengthEngineMock = new Mock<ISpotLengthEngine>();
            _PricingApiClientMock = new Mock<IPricingApiClient>();
            _BackgroundJobClientMock = new Mock<IBackgroundJobClient>();
            _PlanPricingInventoryEngineMock = new Mock<IPlanPricingInventoryEngine>();
            _BroadcastLockingManagerApplicationServiceMock = new Mock<IBroadcastLockingManagerApplicationService>();
        }

        [Test]
        [TestCase(BackgroundJobProcessingStatus.Queued)]
        [TestCase(BackgroundJobProcessingStatus.Processing)]
        public void CanNotQueuePricingJobWhenThereIsOneActive(BackgroundJobProcessingStatus status)
        {
            const string expectedMessage = "The pricing model is already running for the plan";

            var planRepositoryMock = new Mock<IPlanRepository>();
            planRepositoryMock
                .Setup(x => x.GetLatestPricingJob(It.IsAny<int>()))
                .Returns(new PlanPricingJob { Status = status });

            _DataRepositoryFactoryMock
                .Setup(x => x.GetDataRepository<IPlanRepository>())
                .Returns(planRepositoryMock.Object);

            _BroadcastLockingManagerApplicationServiceMock
                .Setup(x => x.GetNotUserBasedLockObjectForKey(It.IsAny<string>()))
                .Returns(new object());

            var service = new PlanPricingService(
                _DataRepositoryFactoryMock.Object,
                _SpotLengthEngineMock.Object,
                _PricingApiClientMock.Object,
                _BackgroundJobClientMock.Object,
                _PlanPricingInventoryEngineMock.Object,
                _BroadcastLockingManagerApplicationServiceMock.Object);

            var exception = Assert.Throws<Exception>(() => service.QueuePricingJob(
                planPricingParametersDto: new PlanPricingParametersDto(), 
                currentDate: new DateTime(2019, 10, 23)));

            Assert.AreEqual(expectedMessage, exception.Message);
        }
    }
}
