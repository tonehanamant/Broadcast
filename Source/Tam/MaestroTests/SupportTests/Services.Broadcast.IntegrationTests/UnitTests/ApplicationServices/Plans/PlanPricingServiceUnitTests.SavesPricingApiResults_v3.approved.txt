[
  {
    "JobId": 1,
    "PlanVersionId": 0,
    "PricingCpm": 0.3569946035699460356994604000,
    "PricingVersion": "3",
    "RequestId": "djj4j4399fmmf1m212",
    "Spots": [
      {
        "ContractMediaWeek": null,
        "Id": 1,
        "Impressions30sec": 1000.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1000.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "AMN",
          "FullName": "AM News",
          "Id": 15,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 1500.0,
        "TotalSpots": 2
      },
      {
        "ContractMediaWeek": null,
        "Id": 1,
        "Impressions30sec": 1000.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1000.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "AMN",
          "FullName": "AM News",
          "Id": 15,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 1500.0,
        "TotalSpots": 2
      },
      {
        "ContractMediaWeek": null,
        "Id": 1,
        "Impressions30sec": 1000.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1000.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "AMN",
          "FullName": "AM News",
          "Id": 15,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 1500.0,
        "TotalSpots": 2
      },
      {
        "ContractMediaWeek": null,
        "Id": 1,
        "Impressions30sec": 1000.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1000.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "AMN",
          "FullName": "AM News",
          "Id": 15,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 1500.0,
        "TotalSpots": 2
      },
      {
        "ContractMediaWeek": null,
        "Id": 1,
        "Impressions30sec": 1000.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1000.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "AMN",
          "FullName": "AM News",
          "Id": 15,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 1500.0,
        "TotalSpots": 2
      },
      {
        "ContractMediaWeek": null,
        "Id": 1,
        "Impressions30sec": 1000.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1000.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "AMN",
          "FullName": "AM News",
          "Id": 15,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 1500.0,
        "TotalSpots": 2
      },
      {
        "ContractMediaWeek": null,
        "Id": 2,
        "Impressions30sec": 1500.0,
        "InventoryMediaWeek": null,
        "SpotFrequencies": [
          {
            "Impressions": 1500.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 1,
            "Spots": 1
          },
          {
            "Impressions": 750.0,
            "SpotCost": 50.0,
            "SpotCostWithMargin": 62.5,
            "SpotLengthId": 2,
            "Spots": 1
          }
        ],
        "StandardDaypart": {
          "Code": "PMN",
          "FullName": "PM News",
          "Id": 16,
          "VpvhCalculationSourceType": 0
        },
        "StationInventoryManifestId": 0,
        "TotalCost": 100.0,
        "TotalCostWithMargin": 125.0,
        "TotalImpressions": 2250.0,
        "TotalSpots": 2
      }
    ]
  }
]